lib/src/async-compiler.ts

```diff
@@ -6,42 +6,174 @@ import {spawn} from 'child_process';
+      throw utils.compilerError('Async compiler has already been disposed.');
+    }
+  }
+
+  /**
+   * Sends a compile request to the child process and returns a Promise that
+   * resolves with the CompileResult. Rejects the promise if there were any
+   * protocol or compilation errors.
+   */
+  private async compileRequestAsync(
+    request: proto.InboundMessage_CompileRequest,
+    importers: ImporterRegistry<'async'>,
+    options?: OptionsWithLegacy<'async'> & {legacy?: boolean}
+  ): Promise<CompileResult> {
+    const functions = new FunctionRegistry(options?.functions);
+
+    const dispatcher = createDispatcher<'async'>(
+      this.compilationId++,
+      this.messageTransformer,
+      {
+        handleImportRequest: request => importers.import(request),
+        handleFileImportRequest: request => importers.fileImport(request),
+        handleCanonicalizeRequest: request => importers.canonicalize(request),
+        handleFunctionCallRequest: request => functions.call(request),
+      }
+    );
+    dispatcher.logEvents$.subscribe(event => handleLogEvent(options, event));
+
+    const compilation = new Promise<proto.OutboundMessage_CompileResponse>(
+      (resolve, reject) =>
+        dispatcher.sendCompileRequest(request, (err, response) => {
+          this.compilations.delete(compilation);
+          if (this.compilations.size === 0) this.compilationId = 1;
+          if (err) {
+            reject(err);
+          } else {
+            resolve(response!);
+          }
+        })
+    );
+    this.compilations.add(compilation);
+
+    return handleCompileResponse(await compilation);
+  }
+
+  /** Initialize resources shared across compilations. */
+  constructor(flag: Symbol | undefined) {
+    if (flag !== initFlag) {
+      throw utils.compilerError(
+        'AsyncCompiler can not be directly constructed. Please use `sass.initAsyncCompiler()` instead.'
+      );
+    }
+    this.stderr$.subscribe(data => process.stderr.write(data));
+    const packetTransformer = new PacketTransformer(this.stdout$, buffer => {
+      this.writeStdin(buffer);
+    });
+    this.messageTransformer = new MessageTransformer(
+      packetTransformer.outboundProtobufs$,
+      packet => packetTransformer.writeInboundProtobuf(packet)
+    );
+  }
+
+  compileAsync(
+    path: string,
+    options?: OptionsWithLegacy<'async'>
+  ): Promise<CompileResult> {
+    this.throwIfDisposed();
+    const importers = new ImporterRegistry(options);
+    return this.compileRequestAsync(
+      newCompilePathRequest(path, importers, options),
+      importers,
+      options
+    );
+  }
+
+  compileStringAsync(
+    source: string,
+    options?: StringOptionsWithLegacy<'async'>
+  ): Promise<CompileResult> {
+    this.throwIfDisposed();
+    const importers = new ImporterRegistry(options);
+    return this.compileRequestAsync(
+      newCompileStringRequest(source, importers, options),
+      importers,
+      options
+    );
+  }
+
+  async dispose() {
```

Comment:
All functions should have declared return types:

```suggestion
  async dispose(): Promise<void> {
```

Reply:
Fixed in bf7cf3c68097e93ce73693a2fa418870ae4aa65b